include/seastar/core/estimated_histogram.hh

```diff
@@ -0,0 +1,385 @@
+            if (_buckets[i]) {
+                elements += _buckets[i];
+                if (elements >= pcount) {
+                    return get_bucket_lower_limit(i);
+                }
+            }
+        }
+        return Max; // overflowed value is in the requested quantile
+    }
+
+    /*!
+     * \brief returns the mean histogram value (average of bucket offsets, weighted by count)
+     * It will return 0 if the histogram is empty.
+     */
+    uint64_t mean() const {
+        uint64_t elements = 0;
+        double sum = 0;
+        for (size_t i = 0; i < NUM_BUCKETS - 1; i++) {
+            elements += _buckets[i];
+            sum += _buckets[i] * get_bucket_lower_limit(i);
+        }
+        return (elements) ?  sum / elements : 0;
+    }
+
+    /*!
+     * \brief returns the number of buckets;
+     */
+    size_t size() const {
+        return NUM_BUCKETS;
+    }
+
+    /*!
+     * \brief returns the total number of values inserted
+     */
+    uint64_t count() const {
+        uint64_t sum = 0L;
+        for (size_t i = 0; i < NUM_BUCKETS; i++) {
+            sum += _buckets[i];
+        }
+        return sum;
+    }
+
+    /*!
+     * \brief multiple all the buckets content in the histogram by a constant
+     */
+    approx_exponential_histogram& operator*=(double v) {
+        for (size_t i = 0; i < NUM_BUCKETS; i++) {
+            _buckets[i] *= v;
+        }
+        return *this;
+    }
+
+    uint64_t& operator[](size_t b) noexcept {
+        return _buckets[b];
+    }
+
+    seastar::metrics::histogram to_metrics_histogram() const noexcept {
+        seastar::metrics::histogram res;
+        res.buckets.resize(size() - 1);
+        uint64_t cummulative_count = 0;
+        res.sample_sum = 0;
+
+        for (size_t i = 0; i < NUM_BUCKETS - 1; i++) {
+            auto& v = res.buckets[i];
+            v.upper_bound = get_bucket_lower_limit(i + 1);
+            cummulative_count += get(i);
+            v.count = cummulative_count;
+            res.sample_sum += get(i) * v.upper_bound;
+        }
+        // The count serves as the infinite bucket
+        res.sample_count = cummulative_count + get(NUM_BUCKETS - 1);
+        res.sample_sum += get(NUM_BUCKETS - 1) * get_bucket_lower_limit(NUM_BUCKETS - 1);
+        return res;
+    }
+};
+
+template<uint64_t Min, uint64_t Max, size_t NumBuckets>
+inline approx_exponential_histogram<Min, Max, NumBuckets> base_estimated_histogram_merge(approx_exponential_histogram<Min, Max, NumBuckets> a, const approx_exponential_histogram<Min, Max, NumBuckets>& b) {
+    return a.merge(b);
+}
+
+/*!
+ * \brief estimated histogram for duration values
+ * time_estimated_histogram is used for short task timing.
+ * It covers the range of 0.5ms to 33s with a precision of 4.
+ *
+ * 512us, 640us, 768us, 896us, 1024us, 1280us, 1536us, 1792us...16s, 20s, 25s, 29s, 33s (33554432us)
+ */
+class time_estimated_histogram : public approx_exponential_histogram<512, 33554432, 4> {
+public:
```

Comment:
Why is time_estimated_histogram choosing these values? Different measurements will need different parameters. Best to keep this a template.

Reply:
It's taken from Scylla As-Is, I took it with as minimal changes as possible, I can link to the discussion when it was added back then.

When I'll do more changes to the histograms mechanism, I'll update it here, but i think we should be consisten between the implementations.

Reply:
@avikivity Are you aware that all histogram code in this PR was copied as-is from the Scylla repository? Focusing the review on years-old code, which @amnonh isn't an author of, seems counterproductive.

Reply:
Taking 7 year old mistakes and enshrining them as an API is not a way to bypass review.

We bypassed review in another way, by making things internal, but copy/paste in general is not an excuse for bad code.